src/MonitorProfiler/MainProfiler/MainProfiler.cpp

```diff
@@ -285,15 +293,28 @@ HRESULT MainProfiler::InitializeCommandServer()
 
 HRESULT MainProfiler::MessageCallback(const IpcMessage& message)
 {
-    m_pLogger->Log(LogLevel::Information, _LS("Message received from client: %d %d"), message.MessageType, message.Parameters);
+    m_pLogger->Log(LogLevel::Debug, _LS("Message received from client %d"), message.Command);
 
-    if (message.MessageType == MessageType::Callstack)
+    switch (message.Command)
     {
-        //Currently we do not have any options for this message
-        return ProcessCallstackMessage();
+        case IpcCommand::Unknown:
+            return E_FAIL;
+        case IpcCommand::Callstack:
+            return ProcessCallstackMessage();
+        default:
+            lock_guard<mutex> lock(g_messageCallbackMutex);
+            if (g_pManagedMessageCallback == nullptr)
+            {
+                return E_FAIL;
+            }
+
+            return g_pManagedMessageCallback(
```

Comment:
is the intent to lock around the message callback itself or just access to g_pManagedMessageCallback variable? For the former, I believe we already sequentially dequeue messages. For the latter, consider std::atomic or similar lockfree calls.